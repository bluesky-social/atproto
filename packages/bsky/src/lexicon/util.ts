/**
 * GENERATED CODE - DO NOT MODIFY
 */
export function isObj(v: unknown): v is Record<string, unknown> {
  return typeof v === 'object' && v !== null
}

export function hasProp<O extends object, K extends PropertyKey>(
  data: O,
  prop: K,
): data is O & Record<K, unknown> {
  return prop in data
}
