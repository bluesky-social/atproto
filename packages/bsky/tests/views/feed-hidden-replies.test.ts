import { TestNetwork, SeedClient } from '@atproto/dev-env'
import AtpAgent from '@atproto/api'

import { ids } from '../../src/lexicon/lexicons'
import { feedHiddenRepliesSeed, Users } from '../seed/feed-hidden-replies'

describe('feed hidden replies', () => {
  let network: TestNetwork
  let agent: AtpAgent
  let pdsAgent: AtpAgent
  let sc: SeedClient
  let users: Users

  beforeAll(async () => {
    network = await TestNetwork.create({
      dbPostgresSchema: 'bsky_tests_feed_hidden_replies',
    })
    agent = network.bsky.getClient()
    pdsAgent = network.pds.getClient()
    sc = network.getSeedClient()

    const result = await feedHiddenRepliesSeed(sc)
    users = result.users

    await network.processAll()
  })

  afterAll(async () => {
    await network.close()
  })

  describe(`notifications`, () => {
    it(`[A] -> [B] : B is hidden`, async () => {
      const A = await sc.post(users.poster.did, `A`)

      await network.processAll()

      const B = await sc.reply(users.replier.did, A.ref, A.ref, `B`)
      const C = await sc.reply(users.replier.did, A.ref, A.ref, `C`)

      await pdsAgent.api.app.bsky.feed.threadgate.create(
        {
          repo: A.ref.uri.host,
          rkey: A.ref.uri.rkey,
        },
        {
          post: A.ref.uriStr,
          createdAt: new Date().toISOString(),
          hiddenReplies: [B.ref.uriStr],
        },
        sc.getHeaders(A.ref.uri.host),
      )

      await network.processAll()

      const {
        data: { notifications },
      } = await agent.api.app.bsky.notification.listNotifications(
        {},
        {
          headers: await network.serviceHeaders(
            users.poster.did,
            ids.AppBskyNotificationListNotifications,
          ),
        },
      )

      const notificationB = notifications.find((item) => {
        return item.uri === B.ref.uriStr
      })
      const notificationC = notifications.find((item) => {
        return item.uri === C.ref.uriStr
      })

      expect(notificationB).toBeUndefined()
      expect(notificationC).toBeDefined()
    })

    it(`[A] -> [B] -> [C] : B is hidden, C results in no notification for A, notification for B`, async () => {
      const A = await sc.post(users.poster.did, `A`)
      await network.processAll()
      const B = await sc.reply(users.replier.did, A.ref, A.ref, `B`)

      await pdsAgent.api.app.bsky.feed.threadgate.create(
        {
          repo: A.ref.uri.host,
          rkey: A.ref.uri.rkey,
        },
        {
          post: A.ref.uriStr,
          createdAt: new Date().toISOString(),
          hiddenReplies: [B.ref.uriStr],
        },
        sc.getHeaders(A.ref.uri.host),
      )

      await network.processAll()

      const C = await sc.reply(users.viewer.did, A.ref, B.ref, `C`)

      await network.processAll()

      const {
        data: { notifications: posterNotifications },
      } = await agent.api.app.bsky.notification.listNotifications(
        {},
        {
          headers: await network.serviceHeaders(
            users.poster.did,
            ids.AppBskyNotificationListNotifications,
          ),
        },
      )

      const posterNotificationB = posterNotifications.find((item) => {
        return item.uri === B.ref.uriStr
      })
      const posterNotificationC = posterNotifications.find((item) => {
        return item.uri === C.ref.uriStr
      })

      expect(posterNotificationB).toBeUndefined()
      expect(posterNotificationC).toBeUndefined()

      const {
        data: { notifications: replierNotifications },
      } = await agent.api.app.bsky.notification.listNotifications(
        {},
        {
          headers: await network.serviceHeaders(
            users.replier.did,
            ids.AppBskyNotificationListNotifications,
          ),
        },
      )

      const replierNotificationC = replierNotifications.find((item) => {
        return item.uri === C.ref.uriStr
      })

      expect(replierNotificationC).toBeDefined()
    })

    it(`[A] -> [B] -> [C] -> [D] : C is hidden, D results in no notification for A or B, notification for C, C exists in B's notifications`, async () => {
      const A = await sc.post(users.poster.did, `A`)
      await network.processAll()
      const B = await sc.reply(users.replier.did, A.ref, A.ref, `B`)
      await network.processAll()
      const C = await sc.reply(users.viewer.did, A.ref, B.ref, `C`)
      await network.processAll()

      const {
        data: { notifications: posterNotificationsBefore },
      } = await agent.api.app.bsky.notification.listNotifications(
        {},
        {
          headers: await network.serviceHeaders(
            users.poster.did,
            ids.AppBskyNotificationListNotifications,
          ),
        },
      )

      const posterNotificationCBefore = posterNotificationsBefore.find(
        (item) => {
          return item.uri === C.ref.uriStr
        },
      )

      expect(posterNotificationCBefore).toBeDefined()

      await pdsAgent.api.app.bsky.feed.threadgate.create(
        {
          repo: A.ref.uri.host,
          rkey: A.ref.uri.rkey,
        },
        {
          post: A.ref.uriStr,
          createdAt: new Date().toISOString(),
          hiddenReplies: [C.ref.uriStr],
        },
        sc.getHeaders(A.ref.uri.host),
      )
      await network.processAll()
      const D = await sc.reply(users.viewer.did, A.ref, C.ref, `D`)
      await network.processAll()

      const {
        data: { notifications: posterNotifications },
      } = await agent.api.app.bsky.notification.listNotifications(
        {},
        {
          headers: await network.serviceHeaders(
            users.poster.did,
            ids.AppBskyNotificationListNotifications,
          ),
        },
      )

      const posterNotificationB = posterNotifications.find((item) => {
        return item.uri === B.ref.uriStr
      })
      const posterNotificationC = posterNotifications.find((item) => {
        return item.uri === C.ref.uriStr
      })
      const posterNotificationD = posterNotifications.find((item) => {
        return item.uri === D.ref.uriStr
      })

      expect(posterNotificationB).toBeDefined()
      expect(posterNotificationC).toBeUndefined() // hidden bc OP
      expect(posterNotificationD).toBeUndefined() // hidden bc no propogation

      const {
        data: { notifications: replierNotifications },
      } = await agent.api.app.bsky.notification.listNotifications(
        {},
        {
          headers: await network.serviceHeaders(
            users.replier.did,
            ids.AppBskyNotificationListNotifications,
          ),
        },
      )

      const replierNotificationC = replierNotifications.find((item) => {
        return item.uri === C.ref.uriStr
      })
      const replierNotificationD = replierNotifications.find((item) => {
        return item.uri === D.ref.uriStr
      })

      expect(replierNotificationC).toBeDefined() // not hidden bc not OP
      expect(replierNotificationD).toBeUndefined() // hidden bc no propogation

      await pdsAgent.api.app.bsky.feed.threadgate.delete(
        {
          repo: A.ref.uri.host,
          rkey: A.ref.uri.rkey,
        },
        sc.getHeaders(A.ref.uri.host),
      )
      await network.processAll()
    })
  })
})
