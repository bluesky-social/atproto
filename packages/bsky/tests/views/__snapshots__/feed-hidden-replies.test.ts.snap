// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`postgates [A] -> [B] : B is hidden 1`] = `
Array [
  Object {
    "post": Object {
      "author": Object {
        "did": "did:plc:2yi6yquajbeucilyxy4dhwr7",
        "handle": "poster.test",
        "labels": Array [],
        "viewer": Object {
          "blockedBy": false,
          "following": "at://did:plc:roxqhd7co2rlpgg25cjw4pbd/app.bsky.graph.follow/3kzmcmgceqk2h",
          "muted": false,
        },
      },
      "cid": "bafyreihhj4lmtv6oy2lkp4us4zhowc3y5pg3wgncnt5xsnaywkradxdu44",
      "indexedAt": "2024-08-13T15:20:02.863Z",
      "labels": Array [],
      "likeCount": 0,
      "record": Object {
        "$type": "app.bsky.feed.post",
        "createdAt": "2024-08-13T15:20:02.863Z",
        "text": "A",
      },
      "replyCount": 1,
      "repostCount": 0,
      "threadgate": Object {
        "cid": "bafyreifq6uewbuj65kjxwl7nxkvldghqujqc6576ezw72vnpg356eyrrqi",
        "lists": Array [],
        "record": Object {
          "$type": "app.bsky.feed.threadgate",
          "createdAt": "2024-08-13T15:20:02.876Z",
          "hiddenReplies": Array [
            "at://did:plc:m5tbkgx36zxyti5me4g2vjad/app.bsky.feed.post/3kzmcmgexqt2h",
          ],
          "post": "at://did:plc:2yi6yquajbeucilyxy4dhwr7/app.bsky.feed.post/3kzmcmgervd2h",
        },
        "uri": "at://did:plc:2yi6yquajbeucilyxy4dhwr7/app.bsky.feed.threadgate/3kzmcmgervd2h",
      },
      "uri": "at://did:plc:2yi6yquajbeucilyxy4dhwr7/app.bsky.feed.post/3kzmcmgervd2h",
      "viewer": Object {
        "embeddingDisabled": false,
        "replyDisabled": false,
        "threadMuted": false,
      },
    },
  },
]
`;

exports[`postgates [A] -> [B] : B is hidden but was reposted 1`] = `
Array [
  Object {
    "post": Object {
      "author": Object {
        "did": "did:plc:m5tbkgx36zxyti5me4g2vjad",
        "handle": "replier.test",
        "labels": Array [],
        "viewer": Object {
          "blockedBy": false,
          "following": "at://did:plc:roxqhd7co2rlpgg25cjw4pbd/app.bsky.graph.follow/3kzmcmgceql2h",
          "muted": false,
        },
      },
      "cid": "bafyreicn2o6cyvl2ak6rxygogb2qxlh4fyijmf4g574b6a2me75m2ev6y4",
      "indexedAt": "2024-08-13T15:20:03.153Z",
      "labels": Array [],
      "likeCount": 0,
      "record": Object {
        "$type": "app.bsky.feed.post",
        "createdAt": "2024-08-13T15:20:03.153Z",
        "reply": Object {
          "parent": Object {
            "cid": "bafyreid3cmkv32rqmjxyyyu72wkdy6wjnbxnjbxlq575tg3bdxoltumbzq",
            "uri": "at://did:plc:2yi6yquajbeucilyxy4dhwr7/app.bsky.feed.post/3kzmcmgnfbt2h",
          },
          "root": Object {
            "cid": "bafyreid3cmkv32rqmjxyyyu72wkdy6wjnbxnjbxlq575tg3bdxoltumbzq",
            "uri": "at://did:plc:2yi6yquajbeucilyxy4dhwr7/app.bsky.feed.post/3kzmcmgnfbt2h",
          },
        },
        "text": "B",
      },
      "replyCount": 0,
      "repostCount": 1,
      "uri": "at://did:plc:m5tbkgx36zxyti5me4g2vjad/app.bsky.feed.post/3kzmcmgnm4l2h",
      "viewer": Object {
        "embeddingDisabled": false,
        "replyDisabled": false,
        "threadMuted": false,
      },
    },
    "reason": Object {
      "$type": "app.bsky.feed.defs#reasonRepost",
      "by": Object {
        "did": "did:plc:kcd2kbxdbnbmbqokuloyy6gw",
        "handle": "reposter.test",
        "labels": Array [],
        "viewer": Object {
          "blockedBy": false,
          "following": "at://did:plc:roxqhd7co2rlpgg25cjw4pbd/app.bsky.graph.follow/3kzmcmgd2a32h",
          "muted": false,
        },
      },
      "indexedAt": "2024-08-13T15:20:03.158Z",
    },
    "reply": Object {
      "parent": Object {
        "$type": "app.bsky.feed.defs#postView",
        "author": Object {
          "did": "did:plc:2yi6yquajbeucilyxy4dhwr7",
          "handle": "poster.test",
          "labels": Array [],
          "viewer": Object {
            "blockedBy": false,
            "following": "at://did:plc:roxqhd7co2rlpgg25cjw4pbd/app.bsky.graph.follow/3kzmcmgceqk2h",
            "muted": false,
          },
        },
        "cid": "bafyreid3cmkv32rqmjxyyyu72wkdy6wjnbxnjbxlq575tg3bdxoltumbzq",
        "indexedAt": "2024-08-13T15:20:03.146Z",
        "labels": Array [],
        "likeCount": 0,
        "record": Object {
          "$type": "app.bsky.feed.post",
          "createdAt": "2024-08-13T15:20:03.146Z",
          "text": "A",
        },
        "replyCount": 1,
        "repostCount": 0,
        "threadgate": Object {
          "cid": "bafyreibahdbdyiu7lhyztnb33brf7iyvwmnrzemwkdbudzztjylmuxzzhe",
          "lists": Array [],
          "record": Object {
            "$type": "app.bsky.feed.threadgate",
            "createdAt": "2024-08-13T15:20:03.164Z",
            "hiddenReplies": Array [
              "at://did:plc:m5tbkgx36zxyti5me4g2vjad/app.bsky.feed.post/3kzmcmgnm4l2h",
            ],
            "post": "at://did:plc:2yi6yquajbeucilyxy4dhwr7/app.bsky.feed.post/3kzmcmgnfbt2h",
          },
          "uri": "at://did:plc:2yi6yquajbeucilyxy4dhwr7/app.bsky.feed.threadgate/3kzmcmgnfbt2h",
        },
        "uri": "at://did:plc:2yi6yquajbeucilyxy4dhwr7/app.bsky.feed.post/3kzmcmgnfbt2h",
        "viewer": Object {
          "embeddingDisabled": false,
          "replyDisabled": false,
          "threadMuted": false,
        },
      },
      "root": Object {
        "$type": "app.bsky.feed.defs#postView",
        "author": Object {
          "did": "did:plc:2yi6yquajbeucilyxy4dhwr7",
          "handle": "poster.test",
          "labels": Array [],
          "viewer": Object {
            "blockedBy": false,
            "following": "at://did:plc:roxqhd7co2rlpgg25cjw4pbd/app.bsky.graph.follow/3kzmcmgceqk2h",
            "muted": false,
          },
        },
        "cid": "bafyreid3cmkv32rqmjxyyyu72wkdy6wjnbxnjbxlq575tg3bdxoltumbzq",
        "indexedAt": "2024-08-13T15:20:03.146Z",
        "labels": Array [],
        "likeCount": 0,
        "record": Object {
          "$type": "app.bsky.feed.post",
          "createdAt": "2024-08-13T15:20:03.146Z",
          "text": "A",
        },
        "replyCount": 1,
        "repostCount": 0,
        "threadgate": Object {
          "cid": "bafyreibahdbdyiu7lhyztnb33brf7iyvwmnrzemwkdbudzztjylmuxzzhe",
          "lists": Array [],
          "record": Object {
            "$type": "app.bsky.feed.threadgate",
            "createdAt": "2024-08-13T15:20:03.164Z",
            "hiddenReplies": Array [
              "at://did:plc:m5tbkgx36zxyti5me4g2vjad/app.bsky.feed.post/3kzmcmgnm4l2h",
            ],
            "post": "at://did:plc:2yi6yquajbeucilyxy4dhwr7/app.bsky.feed.post/3kzmcmgnfbt2h",
          },
          "uri": "at://did:plc:2yi6yquajbeucilyxy4dhwr7/app.bsky.feed.threadgate/3kzmcmgnfbt2h",
        },
        "uri": "at://did:plc:2yi6yquajbeucilyxy4dhwr7/app.bsky.feed.post/3kzmcmgnfbt2h",
        "viewer": Object {
          "embeddingDisabled": false,
          "replyDisabled": false,
          "threadMuted": false,
        },
      },
    },
  },
  Object {
    "post": Object {
      "author": Object {
        "did": "did:plc:2yi6yquajbeucilyxy4dhwr7",
        "handle": "poster.test",
        "labels": Array [],
        "viewer": Object {
          "blockedBy": false,
          "following": "at://did:plc:roxqhd7co2rlpgg25cjw4pbd/app.bsky.graph.follow/3kzmcmgceqk2h",
          "muted": false,
        },
      },
      "cid": "bafyreid3cmkv32rqmjxyyyu72wkdy6wjnbxnjbxlq575tg3bdxoltumbzq",
      "indexedAt": "2024-08-13T15:20:03.146Z",
      "labels": Array [],
      "likeCount": 0,
      "record": Object {
        "$type": "app.bsky.feed.post",
        "createdAt": "2024-08-13T15:20:03.146Z",
        "text": "A",
      },
      "replyCount": 1,
      "repostCount": 0,
      "threadgate": Object {
        "cid": "bafyreibahdbdyiu7lhyztnb33brf7iyvwmnrzemwkdbudzztjylmuxzzhe",
        "lists": Array [],
        "record": Object {
          "$type": "app.bsky.feed.threadgate",
          "createdAt": "2024-08-13T15:20:03.164Z",
          "hiddenReplies": Array [
            "at://did:plc:m5tbkgx36zxyti5me4g2vjad/app.bsky.feed.post/3kzmcmgnm4l2h",
          ],
          "post": "at://did:plc:2yi6yquajbeucilyxy4dhwr7/app.bsky.feed.post/3kzmcmgnfbt2h",
        },
        "uri": "at://did:plc:2yi6yquajbeucilyxy4dhwr7/app.bsky.feed.threadgate/3kzmcmgnfbt2h",
      },
      "uri": "at://did:plc:2yi6yquajbeucilyxy4dhwr7/app.bsky.feed.post/3kzmcmgnfbt2h",
      "viewer": Object {
        "embeddingDisabled": false,
        "replyDisabled": false,
        "threadMuted": false,
      },
    },
  },
  Object {
    "post": Object {
      "author": Object {
        "did": "did:plc:m5tbkgx36zxyti5me4g2vjad",
        "handle": "replier.test",
        "labels": Array [],
        "viewer": Object {
          "blockedBy": false,
          "following": "at://did:plc:roxqhd7co2rlpgg25cjw4pbd/app.bsky.graph.follow/3kzmcmgceql2h",
          "muted": false,
        },
      },
      "cid": "bafyreibc5noxfdq6xoh2fjweks47lcu2ab7k6rspsq2xip2fcuy6s2knjm",
      "indexedAt": "2024-08-13T15:20:02.991Z",
      "labels": Array [],
      "likeCount": 0,
      "record": Object {
        "$type": "app.bsky.feed.post",
        "createdAt": "2024-08-13T15:20:02.991Z",
        "reply": Object {
          "parent": Object {
            "cid": "bafyreib25avjoszxrctxo3ektajkhc75dc3f5y4ucimwx6h5qpdlgbq75i",
            "uri": "at://did:plc:m5tbkgx36zxyti5me4g2vjad/app.bsky.feed.post/3kzmcmgih3d2h",
          },
          "root": Object {
            "cid": "bafyreiecnkcaoupobooub3cfr2eqsc7p2wabv3k7j24xk3csm3kdsd5zkm",
            "uri": "at://did:plc:2yi6yquajbeucilyxy4dhwr7/app.bsky.feed.post/3kzmcmgiaal2h",
          },
        },
        "text": "C",
      },
      "replyCount": 0,
      "repostCount": 0,
      "uri": "at://did:plc:m5tbkgx36zxyti5me4g2vjad/app.bsky.feed.post/3kzmcmginw32h",
      "viewer": Object {
        "embeddingDisabled": false,
        "replyDisabled": false,
        "threadMuted": false,
      },
    },
    "reply": Object {
      "grandparentAuthor": Object {
        "did": "did:plc:2yi6yquajbeucilyxy4dhwr7",
        "handle": "poster.test",
        "labels": Array [],
        "viewer": Object {
          "blockedBy": false,
          "following": "at://did:plc:roxqhd7co2rlpgg25cjw4pbd/app.bsky.graph.follow/3kzmcmgceqk2h",
          "muted": false,
        },
      },
      "parent": Object {
        "$type": "app.bsky.feed.defs#notFoundPost",
        "notFound": true,
        "uri": "at://did:plc:m5tbkgx36zxyti5me4g2vjad/app.bsky.feed.post/3kzmcmgih3d2h",
      },
      "root": Object {
        "$type": "app.bsky.feed.defs#postView",
        "author": Object {
          "did": "did:plc:2yi6yquajbeucilyxy4dhwr7",
          "handle": "poster.test",
          "labels": Array [],
          "viewer": Object {
            "blockedBy": false,
            "following": "at://did:plc:roxqhd7co2rlpgg25cjw4pbd/app.bsky.graph.follow/3kzmcmgceqk2h",
            "muted": false,
          },
        },
        "cid": "bafyreiecnkcaoupobooub3cfr2eqsc7p2wabv3k7j24xk3csm3kdsd5zkm",
        "indexedAt": "2024-08-13T15:20:02.976Z",
        "labels": Array [],
        "likeCount": 0,
        "record": Object {
          "$type": "app.bsky.feed.post",
          "createdAt": "2024-08-13T15:20:02.976Z",
          "text": "A",
        },
        "replyCount": 1,
        "repostCount": 0,
        "threadgate": Object {
          "cid": "bafyreigglk7uyjostikm6tklur74knkxr4zhaxv2agbqvbvmuzhxfz3xne",
          "lists": Array [],
          "record": Object {
            "$type": "app.bsky.feed.threadgate",
            "createdAt": "2024-08-13T15:20:02.997Z",
            "hiddenReplies": Array [
              "at://did:plc:m5tbkgx36zxyti5me4g2vjad/app.bsky.feed.post/3kzmcmgih3d2h",
            ],
            "post": "at://did:plc:2yi6yquajbeucilyxy4dhwr7/app.bsky.feed.post/3kzmcmgiaal2h",
          },
          "uri": "at://did:plc:2yi6yquajbeucilyxy4dhwr7/app.bsky.feed.threadgate/3kzmcmgiaal2h",
        },
        "uri": "at://did:plc:2yi6yquajbeucilyxy4dhwr7/app.bsky.feed.post/3kzmcmgiaal2h",
        "viewer": Object {
          "embeddingDisabled": false,
          "replyDisabled": false,
          "threadMuted": false,
        },
      },
    },
  },
  Object {
    "post": Object {
      "author": Object {
        "did": "did:plc:2yi6yquajbeucilyxy4dhwr7",
        "handle": "poster.test",
        "labels": Array [],
        "viewer": Object {
          "blockedBy": false,
          "following": "at://did:plc:roxqhd7co2rlpgg25cjw4pbd/app.bsky.graph.follow/3kzmcmgceqk2h",
          "muted": false,
        },
      },
      "cid": "bafyreiecnkcaoupobooub3cfr2eqsc7p2wabv3k7j24xk3csm3kdsd5zkm",
      "indexedAt": "2024-08-13T15:20:02.976Z",
      "labels": Array [],
      "likeCount": 0,
      "record": Object {
        "$type": "app.bsky.feed.post",
        "createdAt": "2024-08-13T15:20:02.976Z",
        "text": "A",
      },
      "replyCount": 1,
      "repostCount": 0,
      "threadgate": Object {
        "cid": "bafyreigglk7uyjostikm6tklur74knkxr4zhaxv2agbqvbvmuzhxfz3xne",
        "lists": Array [],
        "record": Object {
          "$type": "app.bsky.feed.threadgate",
          "createdAt": "2024-08-13T15:20:02.997Z",
          "hiddenReplies": Array [
            "at://did:plc:m5tbkgx36zxyti5me4g2vjad/app.bsky.feed.post/3kzmcmgih3d2h",
          ],
          "post": "at://did:plc:2yi6yquajbeucilyxy4dhwr7/app.bsky.feed.post/3kzmcmgiaal2h",
        },
        "uri": "at://did:plc:2yi6yquajbeucilyxy4dhwr7/app.bsky.feed.threadgate/3kzmcmgiaal2h",
      },
      "uri": "at://did:plc:2yi6yquajbeucilyxy4dhwr7/app.bsky.feed.post/3kzmcmgiaal2h",
      "viewer": Object {
        "embeddingDisabled": false,
        "replyDisabled": false,
        "threadMuted": false,
      },
    },
  },
  Object {
    "post": Object {
      "author": Object {
        "did": "did:plc:2yi6yquajbeucilyxy4dhwr7",
        "handle": "poster.test",
        "labels": Array [],
        "viewer": Object {
          "blockedBy": false,
          "following": "at://did:plc:roxqhd7co2rlpgg25cjw4pbd/app.bsky.graph.follow/3kzmcmgceqk2h",
          "muted": false,
        },
      },
      "cid": "bafyreihhj4lmtv6oy2lkp4us4zhowc3y5pg3wgncnt5xsnaywkradxdu44",
      "indexedAt": "2024-08-13T15:20:02.863Z",
      "labels": Array [],
      "likeCount": 0,
      "record": Object {
        "$type": "app.bsky.feed.post",
        "createdAt": "2024-08-13T15:20:02.863Z",
        "text": "A",
      },
      "replyCount": 1,
      "repostCount": 0,
      "threadgate": Object {
        "cid": "bafyreifq6uewbuj65kjxwl7nxkvldghqujqc6576ezw72vnpg356eyrrqi",
        "lists": Array [],
        "record": Object {
          "$type": "app.bsky.feed.threadgate",
          "createdAt": "2024-08-13T15:20:02.876Z",
          "hiddenReplies": Array [
            "at://did:plc:m5tbkgx36zxyti5me4g2vjad/app.bsky.feed.post/3kzmcmgexqt2h",
          ],
          "post": "at://did:plc:2yi6yquajbeucilyxy4dhwr7/app.bsky.feed.post/3kzmcmgervd2h",
        },
        "uri": "at://did:plc:2yi6yquajbeucilyxy4dhwr7/app.bsky.feed.threadgate/3kzmcmgervd2h",
      },
      "uri": "at://did:plc:2yi6yquajbeucilyxy4dhwr7/app.bsky.feed.post/3kzmcmgervd2h",
      "viewer": Object {
        "embeddingDisabled": false,
        "replyDisabled": false,
        "threadMuted": false,
      },
    },
  },
]
`;

exports[`postgates [A] -> [B] -> [C] : B is hidden, C has tombstone on parent 1`] = `
Array [
  Object {
    "post": Object {
      "author": Object {
        "did": "did:plc:m5tbkgx36zxyti5me4g2vjad",
        "handle": "replier.test",
        "labels": Array [],
        "viewer": Object {
          "blockedBy": false,
          "following": "at://did:plc:roxqhd7co2rlpgg25cjw4pbd/app.bsky.graph.follow/3kzmcmgceql2h",
          "muted": false,
        },
      },
      "cid": "bafyreibc5noxfdq6xoh2fjweks47lcu2ab7k6rspsq2xip2fcuy6s2knjm",
      "indexedAt": "2024-08-13T15:20:02.991Z",
      "labels": Array [],
      "likeCount": 0,
      "record": Object {
        "$type": "app.bsky.feed.post",
        "createdAt": "2024-08-13T15:20:02.991Z",
        "reply": Object {
          "parent": Object {
            "cid": "bafyreib25avjoszxrctxo3ektajkhc75dc3f5y4ucimwx6h5qpdlgbq75i",
            "uri": "at://did:plc:m5tbkgx36zxyti5me4g2vjad/app.bsky.feed.post/3kzmcmgih3d2h",
          },
          "root": Object {
            "cid": "bafyreiecnkcaoupobooub3cfr2eqsc7p2wabv3k7j24xk3csm3kdsd5zkm",
            "uri": "at://did:plc:2yi6yquajbeucilyxy4dhwr7/app.bsky.feed.post/3kzmcmgiaal2h",
          },
        },
        "text": "C",
      },
      "replyCount": 0,
      "repostCount": 0,
      "uri": "at://did:plc:m5tbkgx36zxyti5me4g2vjad/app.bsky.feed.post/3kzmcmginw32h",
      "viewer": Object {
        "embeddingDisabled": false,
        "replyDisabled": false,
        "threadMuted": false,
      },
    },
    "reply": Object {
      "grandparentAuthor": Object {
        "did": "did:plc:2yi6yquajbeucilyxy4dhwr7",
        "handle": "poster.test",
        "labels": Array [],
        "viewer": Object {
          "blockedBy": false,
          "following": "at://did:plc:roxqhd7co2rlpgg25cjw4pbd/app.bsky.graph.follow/3kzmcmgceqk2h",
          "muted": false,
        },
      },
      "parent": Object {
        "$type": "app.bsky.feed.defs#notFoundPost",
        "notFound": true,
        "uri": "at://did:plc:m5tbkgx36zxyti5me4g2vjad/app.bsky.feed.post/3kzmcmgih3d2h",
      },
      "root": Object {
        "$type": "app.bsky.feed.defs#postView",
        "author": Object {
          "did": "did:plc:2yi6yquajbeucilyxy4dhwr7",
          "handle": "poster.test",
          "labels": Array [],
          "viewer": Object {
            "blockedBy": false,
            "following": "at://did:plc:roxqhd7co2rlpgg25cjw4pbd/app.bsky.graph.follow/3kzmcmgceqk2h",
            "muted": false,
          },
        },
        "cid": "bafyreiecnkcaoupobooub3cfr2eqsc7p2wabv3k7j24xk3csm3kdsd5zkm",
        "indexedAt": "2024-08-13T15:20:02.976Z",
        "labels": Array [],
        "likeCount": 0,
        "record": Object {
          "$type": "app.bsky.feed.post",
          "createdAt": "2024-08-13T15:20:02.976Z",
          "text": "A",
        },
        "replyCount": 1,
        "repostCount": 0,
        "threadgate": Object {
          "cid": "bafyreigglk7uyjostikm6tklur74knkxr4zhaxv2agbqvbvmuzhxfz3xne",
          "lists": Array [],
          "record": Object {
            "$type": "app.bsky.feed.threadgate",
            "createdAt": "2024-08-13T15:20:02.997Z",
            "hiddenReplies": Array [
              "at://did:plc:m5tbkgx36zxyti5me4g2vjad/app.bsky.feed.post/3kzmcmgih3d2h",
            ],
            "post": "at://did:plc:2yi6yquajbeucilyxy4dhwr7/app.bsky.feed.post/3kzmcmgiaal2h",
          },
          "uri": "at://did:plc:2yi6yquajbeucilyxy4dhwr7/app.bsky.feed.threadgate/3kzmcmgiaal2h",
        },
        "uri": "at://did:plc:2yi6yquajbeucilyxy4dhwr7/app.bsky.feed.post/3kzmcmgiaal2h",
        "viewer": Object {
          "embeddingDisabled": false,
          "replyDisabled": false,
          "threadMuted": false,
        },
      },
    },
  },
  Object {
    "post": Object {
      "author": Object {
        "did": "did:plc:2yi6yquajbeucilyxy4dhwr7",
        "handle": "poster.test",
        "labels": Array [],
        "viewer": Object {
          "blockedBy": false,
          "following": "at://did:plc:roxqhd7co2rlpgg25cjw4pbd/app.bsky.graph.follow/3kzmcmgceqk2h",
          "muted": false,
        },
      },
      "cid": "bafyreiecnkcaoupobooub3cfr2eqsc7p2wabv3k7j24xk3csm3kdsd5zkm",
      "indexedAt": "2024-08-13T15:20:02.976Z",
      "labels": Array [],
      "likeCount": 0,
      "record": Object {
        "$type": "app.bsky.feed.post",
        "createdAt": "2024-08-13T15:20:02.976Z",
        "text": "A",
      },
      "replyCount": 1,
      "repostCount": 0,
      "threadgate": Object {
        "cid": "bafyreigglk7uyjostikm6tklur74knkxr4zhaxv2agbqvbvmuzhxfz3xne",
        "lists": Array [],
        "record": Object {
          "$type": "app.bsky.feed.threadgate",
          "createdAt": "2024-08-13T15:20:02.997Z",
          "hiddenReplies": Array [
            "at://did:plc:m5tbkgx36zxyti5me4g2vjad/app.bsky.feed.post/3kzmcmgih3d2h",
          ],
          "post": "at://did:plc:2yi6yquajbeucilyxy4dhwr7/app.bsky.feed.post/3kzmcmgiaal2h",
        },
        "uri": "at://did:plc:2yi6yquajbeucilyxy4dhwr7/app.bsky.feed.threadgate/3kzmcmgiaal2h",
      },
      "uri": "at://did:plc:2yi6yquajbeucilyxy4dhwr7/app.bsky.feed.post/3kzmcmgiaal2h",
      "viewer": Object {
        "embeddingDisabled": false,
        "replyDisabled": false,
        "threadMuted": false,
      },
    },
  },
  Object {
    "post": Object {
      "author": Object {
        "did": "did:plc:2yi6yquajbeucilyxy4dhwr7",
        "handle": "poster.test",
        "labels": Array [],
        "viewer": Object {
          "blockedBy": false,
          "following": "at://did:plc:roxqhd7co2rlpgg25cjw4pbd/app.bsky.graph.follow/3kzmcmgceqk2h",
          "muted": false,
        },
      },
      "cid": "bafyreihhj4lmtv6oy2lkp4us4zhowc3y5pg3wgncnt5xsnaywkradxdu44",
      "indexedAt": "2024-08-13T15:20:02.863Z",
      "labels": Array [],
      "likeCount": 0,
      "record": Object {
        "$type": "app.bsky.feed.post",
        "createdAt": "2024-08-13T15:20:02.863Z",
        "text": "A",
      },
      "replyCount": 1,
      "repostCount": 0,
      "threadgate": Object {
        "cid": "bafyreifq6uewbuj65kjxwl7nxkvldghqujqc6576ezw72vnpg356eyrrqi",
        "lists": Array [],
        "record": Object {
          "$type": "app.bsky.feed.threadgate",
          "createdAt": "2024-08-13T15:20:02.876Z",
          "hiddenReplies": Array [
            "at://did:plc:m5tbkgx36zxyti5me4g2vjad/app.bsky.feed.post/3kzmcmgexqt2h",
          ],
          "post": "at://did:plc:2yi6yquajbeucilyxy4dhwr7/app.bsky.feed.post/3kzmcmgervd2h",
        },
        "uri": "at://did:plc:2yi6yquajbeucilyxy4dhwr7/app.bsky.feed.threadgate/3kzmcmgervd2h",
      },
      "uri": "at://did:plc:2yi6yquajbeucilyxy4dhwr7/app.bsky.feed.post/3kzmcmgervd2h",
      "viewer": Object {
        "embeddingDisabled": false,
        "replyDisabled": false,
        "threadMuted": false,
      },
    },
  },
]
`;
